src/relo_core.c

```diff
@@ -1472,39 +1485,116 @@ struct btf *bpf_reloc_info_get_btf(struct btf_reloc_info *info) {
 	return info->src_btf;
 }
 
-static int btf_reloc_info_gen(struct btf_reloc_info *info, const struct bpf_core_relo_res *res) {
-	struct bpf_core_spec *targ_spec = (struct bpf_core_spec *) res->targ_spec;
+static void btf_reloc_dump_spec(struct bpf_core_spec *spec)
+{
+	const struct btf_type *t;
+	const struct btf_enum *e;
+	const char *s;
+	__u32 type_id;
+	int i;
+
+	type_id = spec->root_type_id;
+	t = btf__type_by_id(spec->btf, type_id);
+	s = btf__name_by_offset(spec->btf, t->name_off);
+
+	printf("RELOCATION: [%u] %s %s", type_id, btf_kind_str(t), str_is_empty(s) ? "<anon>" : s);
+
+	if (core_relo_is_type_based(spec->relo_kind)) {
+		printf("\n");
+		return;
+	}
+
+	if (core_relo_is_enumval_based(spec->relo_kind)) {
+		t = skip_mods_and_typedefs(spec->btf, type_id, NULL);
+		e = btf_enum(t) + spec->raw_spec[0];
+		s = btf__name_by_offset(spec->btf, e->name_off);
+
+		printf("::%s = %u\n", s, e->val);
+		return;
+	}
+
+	if (core_relo_is_field_based(spec->relo_kind)) {
+		for (i = 0; i < spec->len; i++) {
+			if (spec->spec[i].name)
+				printf(".%s", spec->spec[i].name);
+			else if (i > 0 || spec->spec[i].idx > 0)
+				printf("[%u]", spec->spec[i].idx);
+		}
+
+		printf(" (");
+		for (i = 0; i < spec->raw_len; i++)
+			printf("%s%d", i == 0 ? "" : ":", spec->raw_spec[i]);
+
+		if (spec->bit_offset % 8)
+			printf(" @ offset %u.%u)", spec->bit_offset / 8, spec->bit_offset % 8);
+		else
+			printf(" @ offset %u)", spec->bit_offset / 8);
+
+		printf("\n");
+		return;
+	}
+}
+
+static int btf_reloc_info_gen_field(struct btf_reloc_info *info, struct bpf_core_spec *targ_spec) {
+
 	struct btf *btf = (struct btf *) targ_spec->btf;
 	struct btf_type *btf_type;
 	struct btf_reloc_type *reloc_type;
 
+	btf_reloc_dump_spec(targ_spec);
+
 	btf_type = btf_type_by_id(btf, targ_spec->root_type_id);
 
 	reloc_type = btf_reloc_get_type(info, btf_type);
-
 	if (reloc_type == NULL) {
 		reloc_type = calloc(1, sizeof(struct btf_reloc_type));
-
 		reloc_type->type = btf_type;
 		reloc_type->id = targ_spec->root_type_id;
-
 		btf_reloc_add_type(btf, info, reloc_type);
 	}
 
+	struct btf_array *a;
+	struct btf_type *this;
+
 	// add types for members of parent type (only for struct or union)
 	for (int i = 1; i < targ_spec->raw_len; i++) {
-		if (btf_is_array(btf_type)) {
-			struct btf_array *a = btf_array(btf_type);
-			struct btf_type *array_type = (struct btf_type *)btf__type_by_id(btf, a->type);
-
-			// set the current type to array type and continue processing
-			reloc_type = btf_reloc_get_type(info, array_type);
 
+		if (!targ_spec->spec[i].name)
 			continue;
+		else {
+			while (btf_is_mod(btf_type) || btf_is_typedef(btf_type)) {
+				btf_type = (struct btf_type*) btf__type_by_id(btf, btf_type->type);
+				reloc_type = btf_reloc_get_type(info, btf_type);
```

Comment:
What about doing it on `case BTF_KIND_TYPEDEF`?
Do you know/think there is any case where `btf_is_mod()` would be true?

Reply:
A volatile inside a struct:

[80] STRUCT 'sock_common' size=187 vlen=6
        'skc_family' type_id=81 bits_offset=128
        '(anon)' type_id=83 bits_offset=0
        '(anon)' type_id=86 bits_offset=96
        'skc_state' type_id=98 bits_offset=144
        'skc_v6_rcv_saddr' type_id=100 bits_offset=576
        'skc_v6_daddr' type_id=100 bits_offset=448

type 98 and 118: 

[98] VOLATILE '(anon)' type_id=118
[118] INT 'unsigned char' size=1 bits_offset=0 nr_bits=8 encoding=(none)

It gets added by its own relocation speculation (btf_reloc_add_type) BUT it needs to be skipped when the relocation that contains it as a member is reached.

Unless I did not understand this correctly =o)

Reply:
Its easier to do it before because we need to skip typedef of typedefs as well :\. This loop is actually taken from the skip_mod_and_typedef (or something like it). There is a function with that approach BUT we needed the reloc to be redefined each resolution.. so I wrote it again instead of using it.

This while the while loop solves all the mods/typedefs before moving on into the add type logic.